cpp/src/arrow/compare.cc

```diff
@@ -376,13 +393,46 @@ class EqualsVisitor : public RangeEqualsVisitor {
 
   Status Visit(const IntervalArray& left) override { return ComparePrimitive(left); }
 
+  template <typename ArrayType>
+  bool ValueOffsetsEqual(const ArrayType& left) {
+    const auto& right = static_cast<const ArrayType&>(right_);
+
+    if (left.offset() == 0 && right.offset() == 0) {
+      return left.value_offsets()->Equals(
+          *right.value_offsets(), (left.length() + 1) * sizeof(int32_t));
+    } else {
+      // One of the arrays is sliced; logic is more complicated because the
+      // value offsets are not both 0-based
+      auto left_offsets =
+          reinterpret_cast<const int32_t*>(left.value_offsets()->data()) + left.offset();
+      auto right_offsets =
+          reinterpret_cast<const int32_t*>(right.value_offsets()->data()) +
+          right.offset();
+
+      for (int32_t i = 0; i < left.length() + 1; ++i) {
+        if (left_offsets[i] - left_offsets[0] != right_offsets[i] - right_offsets[0]) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+
   bool CompareBinary(const BinaryArray& left) {
     const auto& right = static_cast<const BinaryArray&>(right_);
-    bool equal_offsets =
-        left.offsets()->Equals(*right.offsets(), (left.length() + 1) * sizeof(int32_t));
+
+    bool equal_offsets = ValueOffsetsEqual<BinaryArray>(left);
     if (!equal_offsets) { return false; }
-    if (!left.data() && !(right.data())) { return true; }
-    return left.data()->Equals(*right.data(), left.raw_offsets()[left.length()]);
+
+    if (left.offset() == 0 && right.offset() == 0) {
+      if (!left.data() && !(right.data())) { return true; }
+      return left.data()->Equals(*right.data(), left.raw_value_offsets()[left.length()]);
+    } else {
+      // Compare the corresponding data range
```

Comment:
If this is a "bug", it was present before this =) I agree that this is incorrect if there is a null value in a slot with non-zero length according to the offsets and data (e.g. a null bitmap "overlay" of some existing data). Let me add a test case for this